src/validation.cpp

```diff
@@ -3542,6 +3542,80 @@ static void NotifyHeaderTip() LOCKS_EXCLUDED(cs_main) {
     }
 }
 
+bool IsAncestor(const CBlockIndex* pindexChain, const CBlockIndex* pindexCheck)
+{
+    if (!pindexChain || !pindexCheck)
+        return false;
+    if (pindexChain->nHeight <= pindexCheck->nHeight)
+        return false;
+
+    return pindexChain->GetAncestor(pindexCheck->nHeight)->GetBlockHash() == pindexCheck->GetBlockHash();
+}
+
+static const unsigned int PRUNE_DEPTH = 120;
+static const unsigned int PRUNE_COUNT = 12;
+
+/*
+** Periodically scan the block index map for stale tips and clean them up.  Note that both
+** chainActive.Contains and IsAncestor are expensive as the block index grows, so we want
+** to take special care to not spend a lot of extra time doing useless work.
+*/
+void PruneStaleBlockIndexes()
+{
+    static int64_t lastPrunedHeight = 0;
+
+    // This isn't going to change while we're processing, so just leave and try again later.
+    if (!pindexBestHeader) return;
+
+    // If mapBlockIndex isn't bloated, don't bother taking the time.
+    if (chainActive.Height() > (mapBlockIndex.size() - PRUNE_COUNT)) {
+        return;
+    }
+
+    LogPrintf("%s: Checking for stale indexes. Block index size=%d; Chain height=%d\n",
+              __func__, mapBlockIndex.size(), chainActive.Height());
+
+    uint32_t irrelevantIndexes = 0;
+    int64_t lowestPrunedHeight = lastPrunedHeight;
+    std::set<uint256> setDelete;
+
+    for (const auto& p : mapBlockIndex) {
+        const CBlockIndex* pindex = p.second;
+        // If we've already checked to this height, don't waste time
+
+        if (pindex->nHeight < lastPrunedHeight)
+            continue;
+
+        // If it's not in the active chain, and not in our best header ancestor list.
+        if (!chainActive.Contains(pindex) && (!IsAncestor(pindexBestHeader, pindex))) {
+            irrelevantIndexes++;
+
+            // if it's also old enough, add it to the prune list.
+            if (pindex->nHeight + PRUNE_DEPTH < chainActive.Height()) {
+                setDelete.emplace(p.first);
+
+                // save the lowest height that we're purging
+                if ((pindex->nHeight < lowestPrunedHeight) || (lowestPrunedHeight <= lastPrunedHeight)) {
+                    lowestPrunedHeight = pindex->nHeight;
+                }
+            }
+        }
+    }
+
+    if (irrelevantIndexes > 0) {
+         LogPrintf("%s: Erasing %d of %d irrelevant indexes.  LastPrunedHeight now %d.\n",
```

Comment:
Seeing this in my debug log, I'm almost thinking it is better not to have it on the regular debug level, but some other. I just want the debug log a bit more cleaned up and seeing this spammed per block can get a bit annoying.

Reply:
At this point I'm not sure I want to suppress the logging.  It doesn't start logging until the bloat exceeds the threshold to start looking for ones to prune.  The reason I don't want to remove the logging yet?  Because this PR is a symptom repair, not root cause.  We still have efficiency issues with the network itself, and therefore still far too many side chains competing.  When those issues start getting unraveled; the number of tips will begin to reduce and we will be able to monitor that progress somewhat by watching the frequency of this log message.

Quite simply; there is going to be PRs coming where I ask you for a log file that includes these two lines before the change, and after the change, to help measure the improvement.

Do you disagree or concur?

Reply:
I would rather over log a bit at this point.